### PR TITLE
docs(Reply): model unified `Reply` object

### LINKED ISSUE #1617: model more-unified handling of `Reply` and `Message` objects
Between #1493 and #1486, we discovered the following asymmetries in how messages (i.e., of different kinds and directions) are handled in the Client:

1. `DraftReply` objects are separate from, rather than states of, `Reply` objects;
2. `Reply` objects are separate from, rather than directions of, `Message` objects (granted, they are in the Server's data model too);
3. replies sent to the Journalist API can be pending, successful, or failed; and
4. messages (including replies) reported by the Journalist API can be not yet downloaded, downloaded, or failed.

This is a stub ticket to track an effort to consider and model to what extent we can resolve these asymmetries into one (`Message`) or two (`Message` and `Reply`) well-defined state machine/s (a) entirely Client-side or (b) with changes also to the Journalist API and its backing data model.

### PR DESCRIPTION
**Pull request by cfm**
# Description

Closes #1617 by implementing a TLA+ module `Reply` that models the following SecureDrop Client components:

- the unified `securedrop_client.db.Reply` object proposed in
  freedomofpress/securedrop-engineering#8; and
- `securedrop_client.queue.RunnableQueue`, with respect to:
    - `securedrop_client.api_jobs.downloads.DownloadReplyJob`;
    - `securedrop_client.api_jobs.uploads.SendReplyJob`; and
    - `securedrop_client.api_jobs.sources.DeleteConversationJob` as it pertains
      to an individual `Reply` object.

Specifically, this module models the scenarios possible under the following
events, according to the parameters tunable in `Reply.cfg`:

1. The Client attempts to send one or more replies.

2. The Client attempts to download one or more replies, any of which may be a
   reply that the Client previously failed to send.

3. The Client attempts to delete one or more replies.

4. The Client learns that one or more replies has been deleted on the server.

# Test Plan

This is our very first model, and it has no implementation side effects.  @eaon, it's up to you, but I suggest the following review criteria:

- [ ] **Comprehensibility:**  If you either (a) step through the commit history and/or (b) read through the diff, does the model make sense?
    - I have documented the model exhaustively, almost [literately], so that someone unfamiliar with TLA+ can read through it to learn _how TLA+ works_ as well as _how this model works_.  Future models can probably be documented more concisely.
- [ ] **Runnability:**  Does `make Reply` behave as expected?
    - I use the TLA+ extension in Visual Studio Code, so this is only my secondary means of testing the model.
- [ ] **Tunability:**  Do the parameters in `Reply.cfg` make sense?  Does tuning them produce the expected output?
- [ ] **Insight:**  Does the module reveal useful properties of the components and behavior it models?

[literately]: https://en.wikipedia.org/wiki/Literate_programming

### CODE REVIEW

**Comment by cfm** [comment]
@eaon, whatever you think of the review criteria I've suggested here, let me know if you'd rather look at it together....

File: docs/tla/Reply.tla [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+    \* This is the set of IDs for which we've processed deletion operations in
+    \* this model run.  It's an implementation detail of the model only.
+    deleting
+
+\* A reply is alive if it (a) exists and (b) is not dead.
+IsAlive(id) ==
+    /\ id \in DOMAIN pool
+    /\ pool[id] \notin DeadReply
+
+\* A reply is alive if it (a) is alive and (b) has not been locally deleted.
```
Comment (gonzalo-bulnes) [inline docs/tla/Reply.tla:125]:
```suggestion
\* A reply is available if it (a) is alive and (b) has not been locally deleted.
```

**Comment by cfm** [comment]
As a perk following a workshop @eaon and I took with him last year, @hwayne has generously given the following review.  There's great feedback here that I'll incorporate soon.

https://gist.github.com/hwayne/198c04547dc126296c3bff2b9c6560ef

Thank you, @hwayne!

**Comment by cfm** [comment]
Revised per <https://gist.github.com/hwayne/198c04547dc126296c3bff2b9c6560ef>.

**Comment by cfm** [comment]
Rebased from `main` for #1640.

File: docs/tla/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+`$MODEL.tla` is documented:
+
+- in its accompanying configuration `$MODEL.cfg`;
+- with a preamble, above the (e.g.) `---- MODULE Reply ----` line;
+- inline; and
+- with any footnotes or concluding commentary below the `====` line.
+
+With `$TLA_BIN` set in your path, you can run `make $MODEL` to run the model and
+render its state graph. (Or you can let the [TLA+ extension] for Visual Studio
+Code/Codium do all this for you.)
```
Comment (eloquence) [inline docs/tla/README.md:11]: I tried the $TLA_BIN patheway - here are the specific downloads I used:
1. Downloaded the latest `tla2tools.jar` from https://github.com/tlaplus/tlaplus/releases
2. exported `TLA_BIN` as "java -jar /full/path/to/tla2tools.jar"
3. Run `make Reply`

This worked and successfully generated the `Reply.png`. :tada:  If these are the correct steps for the non-extension pathway, that might be worth explicitly adding here.

Comment (cfm) [reply docs/tla/README.md:11]: Good idea.  Documented inline in 5875a0c059db1400f094b5231d9341889af5b7f7.

**Comment by cfm** [comment]
Closing: @zenmonkeykstop and I have concluded that we aren't able to do meaningful review of this kind of modeling work right now.

**Pull request closed without merging**